File: card-infra/src/main/java/com/jean202/cardmizer/infra/persistence/jpa/JpaCardPerformancePolicyAdapter.java
package com.jean202.cardmizer.infra.persistence.jpa;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.jean202.cardmizer.common.Money;
import com.jean202.cardmizer.core.domain.BenefitMonthlyCapTier;
import com.jean202.cardmizer.core.domain.BenefitRule;
import com.jean202.cardmizer.core.domain.BenefitType;
import com.jean202.cardmizer.core.domain.CardId;
import com.jean202.cardmizer.core.domain.CardPerformancePolicy;
import com.jean202.cardmizer.core.domain.PerformanceTier;
import com.jean202.cardmizer.core.port.out.LoadCardPerformancePoliciesPort;
import com.jean202.cardmizer.core.port.out.ReplaceCardPerformancePolicyPort;
import com.jean202.cardmizer.core.port.out.SaveCardPerformancePolicyPort;
import java.io.UncheckedIOException;
import java.util.List;
import java.util.Set;
import org.springframework.context.annotation.Primary;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
@Primary
@Transactional
public class JpaCardPerformancePolicyAdapter implements
        LoadCardPerformancePoliciesPort,
        SaveCardPerformancePolicyPort,
        ReplaceCardPerformancePolicyPort {
    private final JpaCardPerformancePolicyRepository repository;
    private final ObjectMapper objectMapper;

    public JpaCardPerformancePolicyAdapter(
            JpaCardPerformancePolicyRepository repository,
            ObjectMapper objectMapper
    ) {
        this.repository = repository;
        this.objectMapper = objectMapper;
    }

    @Override
    @Transactional(readOnly = true)
    public List<CardPerformancePolicy> loadAll() {
        return repository.findAll().stream()
                .map(this::toDomain)
                .toList();
    }

    @Override
    public void save(CardPerformancePolicy cardPerformancePolicy) {
        if (repository.existsById(cardPerformancePolicy.cardId().value())) {
            throw new IllegalArgumentException("Card policy already exists: " + cardPerformancePolicy.cardId().value());
        }
        repository.save(toEntity(cardPerformancePolicy));
    }

    @Override
    public void replace(CardPerformancePolicy cardPerformancePolicy) {
        repository.save(toEntity(cardPerformancePolicy));
    }

    private JpaCardPerformancePolicyEntity toEntity(CardPerformancePolicy cardPerformancePolicy) {
        return new JpaCardPerformancePolicyEntity(
                cardPerformancePolicy.cardId().value(),
                writeJson(cardPerformancePolicy.tiers().stream().map(PerformanceTierDocument::from).toList()),
                writeJson(cardPerformancePolicy.benefitRules().stream().map(BenefitRuleDocument::from).toList())
        );
    }

    private CardPerformancePolicy toDomain(JpaCardPerformancePolicyEntity entity) {
        List<PerformanceTier> tiers = readJson(entity.getTiersJson(), new TypeReference<List<PerformanceTierDocument>>() {
        }).stream()
                .map(PerformanceTierDocument::toDomain)
                .toList();
        List<BenefitRule> benefitRules = readJson(entity.getBenefitRulesJson(), new TypeReference<List<BenefitRuleDocument>>() {
        }).stream()
                .map(BenefitRuleDocument::toDomain)
                .toList();
        return new CardPerformancePolicy(new CardId(entity.getCardId()), tiers, benefitRules);
    }

    private String writeJson(Object value) {
        try {
            return objectMapper.writeValueAsString(value);
        } catch (JsonProcessingException exception) {
            throw new UncheckedIOException("Failed to serialize card performance policy", exception);
        }
    }

    private <T> T readJson(String value, TypeReference<T> typeReference) {
        try {
            return objectMapper.readValue(value, typeReference);
        } catch (JsonProcessingException exception) {
            throw new UncheckedIOException("Failed to deserialize card performance policy", exception);
        }
    }

    private record PerformanceTierDocument(
            String code,
            long targetAmount,
            String benefitSummary
    ) {
        private static PerformanceTierDocument from(PerformanceTier performanceTier) {
            return new PerformanceTierDocument(
                    performanceTier.code(),
                    performanceTier.targetAmount().amount(),
                    performanceTier.benefitSummary()
            );
        }

        private PerformanceTier toDomain() {
            return new PerformanceTier(code, Money.won(targetAmount), benefitSummary);
        }
    }

    private record BenefitMonthlyCapTierDocument(
            long minimumPreviousMonthSpent,
            long monthlyCap
    ) {
        private static BenefitMonthlyCapTierDocument from(BenefitMonthlyCapTier benefitMonthlyCapTier) {
            return new BenefitMonthlyCapTierDocument(
                    benefitMonthlyCapTier.minimumPreviousMonthSpent().amount(),
                    benefitMonthlyCapTier.monthlyCap().amount()
            );
        }

        private BenefitMonthlyCapTier toDomain() {
            return new BenefitMonthlyCapTier(
                    Money.won(minimumPreviousMonthSpent),
                    Money.won(monthlyCap)
            );
        }
    }

    private record BenefitRuleDocument(
            String ruleId,
            String benefitSummary,
            BenefitType benefitType,
            Set<String> merchantCategories,
            Set<String> merchantKeywords,
            Set<String> requiredTags,
            Set<String> excludedTags,
            String exclusiveGroupId,
            String sharedLimitGroupId,
            Integer rateBasisPoints,
            Long fixedBenefitAmount,
            Long minimumPaymentAmount,
            Long perTransactionCap,
            Long minimumPreviousMonthSpent,
            List<BenefitMonthlyCapTierDocument> monthlyCapTiers,
            Long yearlyBenefitCap,
            Integer monthlyCountLimit,
            Integer yearlyCountLimit,
            List<BenefitMonthlyCapTierDocument> sharedMonthlyCapTiers,
            Long sharedYearlyBenefitCap
    ) {
        private static BenefitRuleDocument from(BenefitRule benefitRule) {
            return new BenefitRuleDocument(
                    benefitRule.ruleId(),
                    benefitRule.benefitSummary(),
                    benefitRule.benefitType(),
                    benefitRule.merchantCategories(),
                    benefitRule.merchantKeywords(),
                    benefitRule.requiredTags(),
                    benefitRule.excludedTags(),
                    benefitRule.exclusiveGroupId(),
                    benefitRule.sharedLimitGroupId(),
                    benefitRule.benefitType() == BenefitType.RATE_PERCENT ? benefitRule.rateBasisPoints() : null,
                    benefitRule.benefitType() == BenefitType.FIXED_AMOUNT ? benefitRule.fixedBenefitAmount().amount() : null,
                    benefitRule.minimumPaymentAmount().amount(),
                    benefitRule.perTransactionCap().amount(),
                    benefitRule.minimumPreviousMonthSpent().amount(),
                    benefitRule.monthlyCapTiers().stream().map(BenefitMonthlyCapTierDocument::from).toList(),
                    benefitRule.yearlyBenefitCap().amount(),
                    benefitRule.monthlyCountLimit(),
                    benefitRule.yearlyCountLimit(),
                    benefitRule.sharedMonthlyCapTiers().stream().map(BenefitMonthlyCapTierDocument::from).toList(),
                    benefitRule.sharedYearlyBenefitCap().amount()
            );
        }

        private BenefitRule toDomain() {
            BenefitRule.Builder builder = switch (benefitType) {
                case RATE_PERCENT -> BenefitRule.percentage(ruleId, benefitSummary, 1)
                        .rateBasisPoints(rateBasisPoints == null ? 0 : rateBasisPoints);
                case FIXED_AMOUNT -> BenefitRule.fixedAmount(
                        ruleId,
                        benefitSummary,
                        Money.won(fixedBenefitAmount == null ? 0L : fixedBenefitAmount)
                );
            };

            builder.categories(toArray(merchantCategories));
            builder.merchantKeywords(toArray(merchantKeywords));
            builder.requiredTags(toArray(requiredTags));
            builder.excludedTags(toArray(excludedTags));
            builder.exclusiveGroup(exclusiveGroupId);
            builder.sharedLimitGroup(sharedLimitGroupId);
            builder.minimumPaymentAmount(Money.won(defaultLong(minimumPaymentAmount)));
            builder.perTransactionCap(Money.won(defaultLong(perTransactionCap)));
            builder.minimumPreviousMonthSpent(Money.won(defaultLong(minimumPreviousMonthSpent)));
            builder.yearlyBenefitCap(Money.won(defaultLong(yearlyBenefitCap)));
            builder.monthlyCountLimit(defaultInt(monthlyCountLimit));
            builder.yearlyCountLimit(defaultInt(yearlyCountLimit));
            builder.sharedYearlyBenefitCap(Money.won(defaultLong(sharedYearlyBenefitCap)));

            if (monthlyCapTiers != null && !monthlyCapTiers.isEmpty()) {
                builder.monthlyCapTiers(monthlyCapTiers.stream()
                        .map(BenefitMonthlyCapTierDocument::toDomain)
                        .toArray(BenefitMonthlyCapTier[]::new));
            }
            if (sharedMonthlyCapTiers != null && !sharedMonthlyCapTiers.isEmpty()) {
                builder.sharedMonthlyCapTiers(sharedMonthlyCapTiers.stream()
                        .map(BenefitMonthlyCapTierDocument::toDomain)
                        .toArray(BenefitMonthlyCapTier[]::new));
            }
            return builder.build();
        }

        private String[] toArray(Set<String> values) {
            return values == null ? new String[0] : values.toArray(String[]::new);
        }

        private long defaultLong(Long value) {
            return value == null ? 0L : value;
        }

        private int defaultInt(Integer value) {
            return value == null ? 0 : value;
        }
    }
}

File: card-infra/src/main/kotlin/com/jean202/cardmizer/infra/persistence/jpa/JpaCardPerformancePolicyAdapter.kt
package com.jean202.cardmizer.infra.persistence.jpa

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import com.jean202.cardmizer.common.Money
import com.jean202.cardmizer.core.domain.BenefitMonthlyCapTier
import com.jean202.cardmizer.core.domain.BenefitRule
import com.jean202.cardmizer.core.domain.BenefitType
import com.jean202.cardmizer.core.domain.CardId
import com.jean202.cardmizer.core.domain.CardPerformancePolicy
import com.jean202.cardmizer.core.domain.PerformanceTier
import com.jean202.cardmizer.core.port.out.LoadCardPerformancePoliciesPort
import com.jean202.cardmizer.core.port.out.ReplaceCardPerformancePolicyPort
import com.jean202.cardmizer.core.port.out.SaveCardPerformancePolicyPort
import org.springframework.context.annotation.Primary
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.io.UncheckedIOException

@Component
@Primary
@Transactional
class JpaCardPerformancePolicyAdapter(
    private val repository: JpaCardPerformancePolicyRepository,
    private val objectMapper: ObjectMapper,
) : LoadCardPerformancePoliciesPort, SaveCardPerformancePolicyPort, ReplaceCardPerformancePolicyPort {

    @Transactional(readOnly = true)
    override fun loadAll(): List<CardPerformancePolicy> =
        repository.findAll().map { toDomain(it) }

    override fun save(cardPerformancePolicy: CardPerformancePolicy) {
        if (repository.existsById(cardPerformancePolicy.cardId.value)) {
            throw IllegalArgumentException("Card policy already exists: ${cardPerformancePolicy.cardId.value}")
        }
        repository.save(toEntity(cardPerformancePolicy))
    }

    override fun replace(cardPerformancePolicy: CardPerformancePolicy) {
        repository.save(toEntity(cardPerformancePolicy))
    }

    private fun toEntity(policy: CardPerformancePolicy) = JpaCardPerformancePolicyEntity(
        cardId = policy.cardId.value,
        tiersJson = writeJson(policy.tiers.map { PerformanceTierDocument.from(it) }),
        benefitRulesJson = writeJson(policy.benefitRules.map { BenefitRuleDocument.from(it) }),
    )

    private fun toDomain(entity: JpaCardPerformancePolicyEntity): CardPerformancePolicy {
        val tiers = readJson<List<PerformanceTierDocument>>(
            entity.tiersJson,
            object : TypeReference<List<PerformanceTierDocument>>() {},
        ).map { it.toDomain() }
        val benefitRules = readJson<List<BenefitRuleDocument>>(
            entity.benefitRulesJson,
            object : TypeReference<List<BenefitRuleDocument>>() {},
        ).map { it.toDomain() }
        return CardPerformancePolicy(CardId(entity.cardId), tiers, benefitRules)
    }

    private fun writeJson(value: Any): String =
        try { objectMapper.writeValueAsString(value) }
        catch (e: Exception) { throw UncheckedIOException("Failed to serialize card performance policy", java.io.IOException(e)) }

    private fun <T> readJson(value: String, typeReference: TypeReference<T>): T =
        try { objectMapper.readValue(value, typeReference) }
        catch (e: Exception) { throw UncheckedIOException("Failed to deserialize card performance policy", java.io.IOException(e)) }

    private data class PerformanceTierDocument(
        val code: String,
        val targetAmount: Long,
        val benefitSummary: String,
    ) {
        fun toDomain() = PerformanceTier(code, Money.won(targetAmount), benefitSummary)

        companion object {
            fun from(tier: PerformanceTier) = PerformanceTierDocument(
                code = tier.code,
                targetAmount = tier.targetAmount.amount,
                benefitSummary = tier.benefitSummary,
            )
        }
    }

    private data class BenefitMonthlyCapTierDocument(
        val minimumPreviousMonthSpent: Long,
        val monthlyCap: Long,
    ) {
        fun toDomain() = BenefitMonthlyCapTier(
            Money.won(minimumPreviousMonthSpent),
            Money.won(monthlyCap),
        )

        companion object {
            fun from(tier: BenefitMonthlyCapTier) = BenefitMonthlyCapTierDocument(
                minimumPreviousMonthSpent = tier.minimumPreviousMonthSpent.amount,
                monthlyCap = tier.monthlyCap.amount,
            )
        }
    }

    private data class BenefitRuleDocument(
        val ruleId: String,
        val benefitSummary: String,
        val benefitType: BenefitType,
        val merchantCategories: Set<String>?,
        val merchantKeywords: Set<String>?,
        val requiredTags: Set<String>?,
        val excludedTags: Set<String>?,
        val exclusiveGroupId: String?,
        val sharedLimitGroupId: String?,
        val rateBasisPoints: Int?,
        val fixedBenefitAmount: Long?,
        val minimumPaymentAmount: Long?,
        val perTransactionCap: Long?,
        val minimumPreviousMonthSpent: Long?,
        val monthlyCapTiers: List<BenefitMonthlyCapTierDocument>?,
        val yearlyBenefitCap: Long?,
        val monthlyCountLimit: Int?,
        val yearlyCountLimit: Int?,
        val sharedMonthlyCapTiers: List<BenefitMonthlyCapTierDocument>?,
        val sharedYearlyBenefitCap: Long?,
    ) {
        fun toDomain() = BenefitRule(
            ruleId = ruleId,
            benefitSummary = benefitSummary,
            benefitType = benefitType,
            merchantCategories = merchantCategories ?: emptySet(),
            merchantKeywords = merchantKeywords ?: emptySet(),
            requiredTags = requiredTags ?: emptySet(),
            excludedTags = excludedTags ?: emptySet(),
            exclusiveGroupId = exclusiveGroupId,
            sharedLimitGroupId = sharedLimitGroupId,
            rateBasisPoints = rateBasisPoints ?: 0,
            fixedBenefitAmount = Money.won(fixedBenefitAmount ?: 0L),
            minimumPaymentAmount = Money.won(minimumPaymentAmount ?: 0L),
            perTransactionCap = Money.won(perTransactionCap ?: 0L),
            minimumPreviousMonthSpent = Money.won(minimumPreviousMonthSpent ?: 0L),
            monthlyCapTiers = monthlyCapTiers?.map { it.toDomain() } ?: emptyList(),
            yearlyBenefitCap = Money.won(yearlyBenefitCap ?: 0L),
            monthlyCountLimit = monthlyCountLimit ?: 0,
            yearlyCountLimit = yearlyCountLimit ?: 0,
            sharedMonthlyCapTiers = sharedMonthlyCapTiers?.map { it.toDomain() } ?: emptyList(),
            sharedYearlyBenefitCap = Money.won(sharedYearlyBenefitCap ?: 0L),
        )

        companion object {
            fun from(rule: BenefitRule) = BenefitRuleDocument(
                ruleId = rule.ruleId,
                benefitSummary = rule.benefitSummary,
                benefitType = rule.benefitType,
                merchantCategories = rule.merchantCategories,
                merchantKeywords = rule.merchantKeywords,
                requiredTags = rule.requiredTags,
                excludedTags = rule.excludedTags,
                exclusiveGroupId = rule.exclusiveGroupId,
                sharedLimitGroupId = rule.sharedLimitGroupId,
                rateBasisPoints = if (rule.benefitType == BenefitType.RATE_PERCENT) rule.rateBasisPoints else null,
                fixedBenefitAmount = if (rule.benefitType == BenefitType.FIXED_AMOUNT) rule.fixedBenefitAmount.amount else null,
                minimumPaymentAmount = rule.minimumPaymentAmount.amount,
                perTransactionCap = rule.perTransactionCap.amount,
                minimumPreviousMonthSpent = rule.minimumPreviousMonthSpent.amount,
                monthlyCapTiers = rule.monthlyCapTiers.map { BenefitMonthlyCapTierDocument.from(it) },
                yearlyBenefitCap = rule.yearlyBenefitCap.amount,
                monthlyCountLimit = rule.monthlyCountLimit,
                yearlyCountLimit = rule.yearlyCountLimit,
                sharedMonthlyCapTiers = rule.sharedMonthlyCapTiers.map { BenefitMonthlyCapTierDocument.from(it) },
                sharedYearlyBenefitCap = rule.sharedYearlyBenefitCap.amount,
            )
        }
    }
}
